scrapy/pqueues.py

```diff
@@ -115,29 +132,66 @@ def qfactory(self, key: int) -> QueueProtocol:
             self.key + "/" + str(key),
         )
 
+    def _sqfactory(self, key: int) -> QueueProtocol:
+        assert self._start_queue_cls is not None
+        return build_from_crawler(
+            self._start_queue_cls,
+            self.crawler,
+            f"{self.key}/{key}s",
+        )
+
     def priority(self, request: Request) -> int:
         return -request.priority
 
     def push(self, request: Request) -> None:
         priority = self.priority(request)
-        if priority not in self.queues:
-            self.queues[priority] = self.qfactory(priority)
-        q = self.queues[priority]
+        is_start_request = request.meta.get("is_start_request", False)
+        if is_start_request and self._start_queue_cls:
+            if priority not in self._start_queues:
+                self._start_queues[priority] = self._sqfactory(priority)
+            q = self._start_queues[priority]
+        else:
+            if priority not in self.queues:
+                self.queues[priority] = self.qfactory(priority)
+            q = self.queues[priority]
         q.push(request)  # this may fail (eg. serialization error)
         if self.curprio is None or priority < self.curprio:
             self.curprio = priority
 
     def pop(self) -> Request | None:
-        if self.curprio is None:
-            return None
-        q = self.queues[self.curprio]
-        m = q.pop()
-        if not q:
-            del self.queues[self.curprio]
-            q.close()
-            prios = [p for p, q in self.queues.items() if q]
-            self.curprio = min(prios) if prios else None
-        return m
+        while self.curprio is not None:
+            if self._start_queues:
+                try:
+                    q = self._start_queues[self.curprio]
+                except KeyError:
+                    pass
```

Comment:
Why is _update_curprio not called here? It seems `self._start_queues` affects the curprio.
If it should be added, then it probably make sense to have a small `_pop` helper with the popping logic, to avoid duplication.

Reply:
The logic is to call `self._update_curprio()` when *both* queues raise a `KeyError`. Here it is not called because self.curprio may still exist in `self.queues` below, and `self._update_curprio()` only needs to be changed once you know that neither queue has the current `curprio`.

Reply:
I see, makes sense!